setup.py

```diff
@@ -23,486 +23,568 @@
+                    print(
+                        f"\n\t\t -- autogen targets for {cat} are unchanged; "
+                        "skipping..."
+                    )
+                    continue
 
-            if not path.isdir(current_dir):
                 print(
-                    f" -- {cat} did not have any previously generated targets; direct copy."
+                    f" -- autogen targets for {cat} are different; "
+                    "replacing with new ones."
                 )
-                shutil.move(target_dir, current_dir)
-                continue
 
-            if autogen_directories_match(current_dir, target_dir):
-                print(f" -- autogen targets for {cat} are unchanged; skipping...")
-                continue
-
-            print(
-                f" -- autogen targets for {cat} are different, replacing with new ones."
-            )
-            shutil.rmtree(current_dir)
-            shutil.move(target_dir, current_dir)
+                shutil.rmtree(cur)
+                shutil.move(tgt, cur)
+
+# Try to autodetect CUDA if user hasn't specified.
+_autodetect_cuda = \
+        env['CUDA_ARCH'] is None \
+        and env['TORCH_HAS_CUDA'] \
+        and not env['DISABLE_LIBNATTEN']
+
+if _autodetect_cuda:
+    env['CUDA_ARCH'] = autodetect_cuda()
+else:
+    print(
+        "Building without CUDA support (libnatten). Either DISABLE_LIBNATTEN "
+        f"was set (DISABLE_LIBNATTEN={env['DISABLE_LIBNATTEN']}) or we could not "
+        f"detect a PyTorch build with CUDA ({torch.cuda.is_available()=}"
+    )
+    env['DISABLE_LIBNATTEN'] = True
+
+natten_version = set_natten_version(
+    *list(map(env.get, ["NATTEN_IS_BUILDING_DIST",
+                        "TORCH_HAS_CUDA",
+                        "TORCH_VER",
+                        "TORCH_CUDA_TAG",
+                        "CWD",
+                       ]
+             )
+         )
+    )
 
 
 class BuildExtension(build_ext):
     def build_extension(self, ext):
-        if BUILD_WITH_CUDA:
-
-            print("Preparing to build LIBNATTEN")
+        if not env['DISABLE_LIBNATTEN']:
             this_dir = path.dirname(path.abspath(__file__))
-            cmake_lists_dir = path.join(this_dir, "csrc")
-            autogen_dir = path.join(cmake_lists_dir, "autogen")
-            scripts_dir = path.join(this_dir, "scripts")
+            cmake_dir = path.join(this_dir, "csrc")
+            autogen_dir = path.join(cmake_dir, "autogen")
+            script_dir = path.join(this_dir, "scripts")
+            os.makedirs(cmake_dir, exist_ok=True)
+            os.makedirs(autogen_dir, exist_ok=True)
 
-            # Hack so that we can build somewhere other than /tmp in development mode.
-            # Also because we want CMake to build everything elsewhere, otherwise pypi will package
-            # build files.
-            build_dir = self.build_lib if NATTEN_BUILD_DIR is None else NATTEN_BUILD_DIR
+            if env['BUILD_DIR'] is not None:
```

Comment:
Note: Current local stage has a context manager here. Commenting to remember to discuss. `self.build_lib` never actually gets called?